packages/api/src/middleware/validation/checkAnimalOrBatch.js

```diff
@@ -13,81 +13,622 @@
+    defaultBreedId = animalOrBatchRecord.default_breed_id;
+  }
+  if (!defaultTypeId && animalOrBatchRecord) {
+    defaultTypeId = animalOrBatchRecord.default_type_id;
+  }
+
+  if (defaultTypeId && defaultBreedId) {
+    checkIdIsNumber(defaultBreedId);
+    const defaultBreed = await DefaultAnimalBreedModel.query().findById(defaultBreedId);
+    if (!defaultBreed) {
+      throw customError('Default breed does not exist');
+    }
+    if (defaultBreed.default_type_id !== defaultTypeId) {
+      throw customError('Breed does not match type');
+    }
+  } else if (!defaultTypeId) {
+    throw customError('Default breed must use default type');
+  }
+};
+
+const checkCustomBreedMatchesType = (
+  animalOrBatch,
+  animalOrBatchRecord,
+  customBreed,
+  default_type_id,
+  custom_type_id,
+) => {
+  // customBreed exists at this point
+  let defaultTypeId = default_type_id;
+  let customTypeId = custom_type_id;
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+
+  // If not editing type, check record type
+  if (!oneExists(typeKeyOptions, animalOrBatch) && animalOrBatchRecord) {
+    defaultTypeId = animalOrBatchRecord.default_type_id;
+    customTypeId = animalOrBatchRecord.custom_type_id;
+  }
+
+  // Custom breed does not match type if defaultId OR customTypeId does not match
+  if (
+    (customBreed.default_type_id && customBreed.default_type_id !== defaultTypeId) ||
+    (customBreed.custom_type_id && customBreed.custom_type_id !== customTypeId)
+  ) {
+    throw customError('Breed does not match type');
+  }
+};
+
+const checkAnimalBreed = async (
+  animalOrBatch,
+  farm_id,
+  animalOrBatchRecord = undefined,
+  creating = true,
+) => {
+  const {
+    default_breed_id,
+    custom_breed_id,
+    breed_name,
+    default_type_id,
+    custom_type_id,
+    type_name,
+  } = animalOrBatch;
+  const breedKeyOptions = ['default_breed_id', 'custom_breed_id', 'breed_name'];
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+
+  // Check if breed is present
+  if (
+    (creating && oneExists(breedKeyOptions, animalOrBatch)) ||
+    oneTruthy([default_breed_id, custom_breed_id, breed_name])
+  ) {
+    checkExactlyOneIsProvided(
+      [default_breed_id, custom_breed_id, breed_name],
+      'default_breed_id, custom_breed_id, or breed_name',
+    );
+  }
+  // Check if breed is present
+  if (!creating && oneExists(breedKeyOptions, animalOrBatch)) {
+    // Overwrite with null in db if editing
+    setFalsyValuesToNull(breedKeyOptions, animalOrBatch);
+  }
+
+  if (
+    oneExists(breedKeyOptions, animalOrBatch) &&
+    !oneTruthy([default_breed_id, custom_breed_id, breed_name])
+  ) {
+    // do nothing if nulling breed
+  } else {
+    // Check if default breed or default type is present
+    if (
+      (oneExists(breedKeyOptions, animalOrBatch) && default_breed_id) ||
+      (oneExists(typeKeyOptions, animalOrBatch) && default_type_id)
```

Comment:
Let me double check this! oneExists is used to discriminatebetween nulls and undefined -- its also quite possible I copied or changed existing logic and did not return to double check since it was harmless.

Reply:
The fewer `oneExists` and `oneTruthy` checks, the better for reviewers!

Reply:
💯

Reply:
I think I have removed as many as possible in the new branch: https://github.com/LiteFarmOrg/LiteFarm/pull/3510/commits/b643951b78fd41b2df38b2473b843313474d0e92